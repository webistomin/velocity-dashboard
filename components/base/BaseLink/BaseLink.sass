+b('base-link')
  font: var(--text-link)
  position: relative

  &::after
    content: ''
    width: 0
    height: 1px
    position: absolute
    left: 0
    bottom: rem(-4px)
    background-color: currentColor
    transition: width 0.2s ease-out

  +m('theme')
    +v('default')
      color: var(--color-primary)

      &:hover::after
        width: 100%

    +v('gray')
      color: var(--color-gray-3)
      transition: color 0.2s ease-out

      &::after
        display: none

      &:hover
        color: var(--color-primary)

