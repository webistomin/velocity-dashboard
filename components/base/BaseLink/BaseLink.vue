<template lang="pug">
  component(
    v-bind='getTag'
    :class='[`base-link_theme_${theme}`, `${isButton ? "btn" : ""}`]'
    @click='onClick'
    ).base-link.link
    span.base-link__content
      slot
</template>

<script lang="ts">
  import Vue from 'vue';
  import Component from 'vue-class-component';

  export interface IBaseLinkDownload {
    is: string;
    href: string;
    download: string;
  }

  export interface IBaseLinkExternal {
    is: string;
    href: string;
    target: string;
    rel: string;
  }

  export interface IBaseLinkRoute {
    is: string;
    to: string;
  }

  export interface IBaseLinkButton {
    is: string;
    type: string;
  }

  const BaseLinkProps = Vue.extend({
    props: {
      to: {
        type: String,
        default: '',
      },
      theme: {
        type: String,
        default: 'default',
      },
      isButton: {
        type: Boolean,
        default: false,
      },
      type: {
        type: String,
        default: 'button',
      },
    },
  });

  @Component({
    name: 'BaseLink',
  })
  export default class BaseLink extends BaseLinkProps {
    public onClick(): void {
      this.$emit('click');
    }

    public get getTag(): IBaseLinkDownload | IBaseLinkExternal | IBaseLinkRoute | IBaseLinkButton {
      const isButton = this.isButton;

      if (isButton) {
        return {
          is: 'button',
          type: this.type,
        };
      }

      const url = this.to;
      if (url.match(/([a-zA-Z0-9\s_\\.\-():])+(.doc|.docx|.pdf|.zip|.rar|.7z|xlsx)$/)) {
        return {
          is: 'a',
          href: url,
          download: '',
        };
      } else if (url.match(/^(http(s)?|ftp):\/\//)) {
        return {
          is: 'a',
          href: url,
          target: '_blank',
          rel: 'noopener noreferrer nofollow',
        };
      }

      return {
        is: 'nuxt-link',
        to: url,
      };
    }
  }
</script>

<style lang="sass">
  @import "BaseLink"
</style>
