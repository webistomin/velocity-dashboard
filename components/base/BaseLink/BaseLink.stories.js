import { text, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import BaseLink from './BaseLink';

export default {
  title: 'BaseLink',
  parameters: {
    options: {
      showPanel: true,
      selectedPanel: 'storybook/knobs/panel',
    },
  },
};

export const Default = () => ({
  components: { BaseLink },
  props: {
    to: {
      type: String,
      default: text('to', '/home'),
    },
    type: {
      type: String,
      default: text('type', 'button'),
    },
    isButton: {
      type: Boolean,
      default: boolean('isButton', false),
    },
    theme: {
      type: String,
      default: text('theme', 'default'),
    },
  },
  methods: {
    onClick: action('emit click'),
  },
  template: `<base-link
    :to='to'
    :theme='theme'
    :isButton='isButton'
    :type='type'
    @click='onClick'
  >{{isButton ? 'Button' : 'Link'}}</base-link>`,
});
