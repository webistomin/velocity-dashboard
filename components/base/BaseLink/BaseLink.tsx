import { VueComponent } from 'types/vue-components';
import { Prop, Component, Emit } from 'nuxt-property-decorator';
import { ButtonTypes } from 'types/common/button-types';

import './BaseLink.sass';

export type IBaseLinkThemes = 'default' | 'gray';

export interface IBaseLinkProps {
  theme?: IBaseLinkThemes;
  onClick?: () => void;
}

export interface IBaseLinkPropsWithLink extends IBaseLinkProps {
  to: string;
  isButton?: never;
  type?: never;
}

export interface IBaseLinkPropsWithButton extends IBaseLinkProps {
  to?: never;
  isButton: boolean;
  type?: ButtonTypes;
}

export type IConditionalBaseLinkProps = IBaseLinkPropsWithLink | IBaseLinkPropsWithButton;

@Component({
  name: 'BaseLink',
})
export default class BaseLink extends VueComponent<IConditionalBaseLinkProps> {
  @Prop({ default: 'default' })
  private theme!: IBaseLinkProps['theme'];

  @Prop()
  private to!: IBaseLinkPropsWithLink['to'];

  @Prop()
  private isButton!: IBaseLinkPropsWithButton['isButton'];

  @Prop({ default: 'button' })
  private type!: IBaseLinkPropsWithButton['type'];

  @Emit('click')
  public onClick(): void {}

  public render() {
    const linkComponentDefaultProps = {
      theme: this.theme,
      class: `link base-link base-link_theme_${this.theme} ${this.isButton ? 'btn' : ''}`,
      on: {
        click: this.onClick,
      },
    };
    const linkComponentContent = <span className='base-link__content'>{this.$slots.default}</span>;
    const isButton = this.isButton;
    const url = this.to;

    // Button link
    if (isButton) {
      return (
        <button {...linkComponentDefaultProps} type={this.type}>
          {linkComponentContent}
        </button>
      );
    }

    // Download link
    if (url.match(/([a-zA-Z0-9\s_\\.\-():])+(.doc|.docx|.pdf|.zip|.rar|.7z|xlsx)$/)) {
      return (
        <a {...linkComponentDefaultProps} href={url} download>
          {linkComponentContent}
        </a>
      );
      // External Link
    } else if (url.match(/^(http(s)?|ftp):\/\//)) {
      return (
        <a {...linkComponentDefaultProps} href={url} target='_blank' rel='noopener noreferrer nofollow'>
          {linkComponentContent}
        </a>
      );
    }

    // Route link
    return (
      <nuxt-link {...linkComponentDefaultProps} to={url}>
        {linkComponentContent}
      </nuxt-link>
    );
  }
}
