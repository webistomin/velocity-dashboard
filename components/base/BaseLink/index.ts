import BaseLink from './BaseLink.vue';

export default BaseLink;
