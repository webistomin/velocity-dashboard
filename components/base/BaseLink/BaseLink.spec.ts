import { mount } from '@vue/test-utils';
import BaseLink from './BaseLink';

describe('BaseLink', () => {
  const factory = (propsData = {}, options = {}) => {
    return mount(BaseLink, {
      propsData: {
        ...propsData,
      },
      ...options,
      stubs: ['nuxt-link'],
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory({
      to: '/',
    }).html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Render as button if "isButton" prop is provided', () => {
    const wrapper = factory({
      to: '/',
      isButton: true,
    });
    const btn = wrapper.find('button');
    expect(btn.exists()).toBe(true);
  });

  it('Render as download link if url match file RegExp', () => {
    const wrapper = factory({
      to: '/test.doc',
    });
    const link = wrapper.find('a');
    expect(link.attributes().download).toBe('true');
  });

  it('Render as external link if url match external RegExp', () => {
    const wrapper = factory({
      to: 'https://www.google.com/',
    });
    const link = wrapper.find('a');
    expect(link.attributes().target).toBe('_blank');
  });

  it('Emit "click" event', () => {
    const wrapper = factory({
      isButton: true,
      to: '/',
    });
    wrapper.vm.onClick();
    expect(wrapper.emitted().click);
  });
});
