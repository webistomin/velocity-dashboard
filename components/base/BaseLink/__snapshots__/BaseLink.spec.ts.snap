// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseLink Matches snapshot 1`] = `"<nuxt-link-stub to=\\"/\\" class=\\"link base-link base-link_theme_default \\"><span class=\\"base-link__content\\"></span></nuxt-link-stub>"`;
